content/blog/2020-11-04-unified-theory-gc/index.md

```diff
@@ -0,0 +1,176 @@
+the two not only make sense in principle but in fact really are already in use
+in practice. Here is a list of the the garbage collection strategies in *rough*
+order from most mark-n-sweepy to most-reference-county:
+
+* **T: Tracing.** This is the classic tracing algorithm. The algorithm must
+  traverse the link graph of the heap to 'mark' objects live. The authors point
+  out that tail recursion can reduce the space cost of the traversal when singly
+  linked structures are present.
+
+* **GT: Tracing Generational Collection.** This form of collection uses a
+  "nursery" to improve the time spent on collection. The idea is that recently
+  allocated objects are more likely to be freed earlier, so we can spend our
+  garbage-collection time more efficiently by looking at newer objects more
+  often.
+
+* **Older First.** (Sketched only). In this approach, the oldest generation
+  is traced first instead of the nursery.
+
+* **GCH: Generational with Counted Heap.** Run normal tracing of the nursery,
+  but do reference counting on the mature remaining space of the heap.
+
+* **Train Algorithm.** (Presented but not analyzed). While this approach was
+  definitely interesting, and they drive home that it fits into their hybrid
+  model (car ordering provides implicit reference counts), I must admit I was not
+  able to glean much intuition about why the approach would be worth
+  implementing.
+
+* **CC: Reference Counting, Trial Deletion.** This is mostly reference counting,
+  but with some tracing of certain "candidates" mixed in to detect cycles (i.e.
+  check if a node's reference count is artificially high due to circular
+  references; evidently the case we care about is when a reference count is
+  non-zero purely due to circular references, in which case its memory should be
+  reclaimed).
+
+* **CD: Deferred Reference Counting.** Do reference counting, but save overhead
+  by not counting mutations of root references. Add items
+  to a Zero Count Table (ZCT) when they are decremented to zero. References from the
+  "root set" (a concept borrowed from tracing meaning "references into the heap
+  from outside of the heap") can keep ZCT entries alive.
+
+* **CT: Reference Counting, Tracing Backup.** Use reference counting without
+  attempting to detect cycles. Occasionally do a full tracing collection to
+  clean up cyclic garbage.
+
+* **C: Reference Counting.** This is the classic reference counting algorithm. 
+  They give a really slick way to reduce space
+  overhead for traversing the links by using the space guaranteed to exist in
+  the dead objects themselves.
+
+The paper concludes with a mathematical analysis of the costs of the
+above collection strategies. The authors point out that the analyses are not 
+big-O, so these costs are actually "real", up to the accuracy of the constants
+for a particular system (and I guess also the constancy of those constants?).
```

Comment:
Hmm, not entirely sure what you mean by this parenthetical—are you suggesting that it's not entirely certain that these factors are actually constants in all settings? That would be reasonable!